include/RED4ext/Containers/DynArray.hpp

```diff
@@ -0,0 +1,518 @@
+    {
+        if (Capacity() >= aCount)
+            return;
+
+        auto newCapacity = CalculateGrowth(aCount);
+        SetCapacity(newCapacity);
+    }
+
+    void ShrinkToFit()
+    {
+        if (Capacity() > Size())
+            SetCapacity(Size());
+    }
+
+    [[nodiscard]] Memory::IAllocator* GetAllocator() const
+    {
+        if (Capacity() == 0)
+        {
+            // Case 1: Allocator is stored instead of entries pointer
+            // It's only 8 bytes for VFT so it fits in a pointer
+            return reinterpret_cast<Memory::IAllocator*>(const_cast<Pointer*>(&entries));
+        }
+        else
+        {
+            // Case 2: Allocator is stored at the end of entries buffer (aligned)
+            auto allocatorPtr = AlignUp(reinterpret_cast<size_t>(&entries[Capacity()]), sizeof(void*));
+            return reinterpret_cast<Memory::IAllocator*>(allocatorPtr);
+        }
+    }
+
+    [[nodiscard]] constexpr Reference Front()
+    {
+        assert(!Empty());
+        return Data()[0];
+    }
+
+    [[nodiscard]] constexpr ConstReference Front() const
+    {
+        assert(!Empty());
+        return Data()[0];
+    }
+
+    [[nodiscard]] constexpr Reference Back()
+    {
+        assert(!Empty());
+        return Data()[Size() - 1];
+    }
+
+    [[nodiscard]] constexpr ConstReference Back() const
+    {
+        assert(!Empty());
+        return Data()[Size() - 1];
+    }
+
+    [[nodiscard]] constexpr Iterator Begin() noexcept
+    {
+        return entries;
+    }
+
+    [[nodiscard]] constexpr ConstIterator Begin() const noexcept
+    {
+        return entries;
+    }
+
+    [[nodiscard]] constexpr ReverseIterator RBegin() noexcept
+    {
+        return ReverseIterator(Begin());
+    }
+
+    [[nodiscard]] constexpr ConstReverseIterator RBegin() const noexcept
+    {
+        return ConstReverseIterator(Begin());
+    }
+
+    [[nodiscard]] constexpr Iterator End() noexcept
+    {
+        return Iterator(entries + size);
+    }
+
+    [[nodiscard]] constexpr ConstIterator End() const noexcept
+    {
+        return ConstIterator(entries + size);
+    }
+
+    [[nodiscard]] constexpr ReverseIterator REnd() noexcept
+    {
+        return ReverseIterator(End());
+    }
+
+    [[nodiscard]] constexpr ConstReverseIterator REnd() const noexcept
+    {
+        return ConstReverseIterator(End());
+    }
+
+    [[nodiscard]] constexpr bool Empty() const noexcept
+    {
+        return Size() == 0;
+    }
+
+    [[nodiscard]] constexpr Pointer Data() noexcept
+    {
+        return entries;
+    }
+
+    [[nodiscard]] constexpr ConstPointer Data() const noexcept
+    {
+        return entries;
+    }
+
+    [[nodiscard]] constexpr SizeType Capacity() const noexcept
+    {
+        return capacity;
+    }
+
+    [[nodiscard]] constexpr SizeType MaxSize() const noexcept
+    {
+#undef max // preprocessor thinks max is a macro under NOMINMAX
+        return std::numeric_limits<SizeType>::max();
```

Comment:
I think the correct limit is `std::min(std::numeric_limits<SizeType>::max(), std::numeric_limits<uint32_t>::max() / sizeof(ValueType))` because allocator interface only allows us to assign up to `std::numeric_limits<uint32_t>::max()` bytes.
This can be `static constexpr`, and this method returns it.